monai/data/dataset.py

```diff
@@ -850,21 +850,20 @@ def _load_cache_item(self, idx: int):
         return item
 
     def _transform(self, index: int):
-        index_: Any = index
         if self.hash_as_key:
             key = self.hash_func(self.data[index])
-            if key in self._cache:
+            if key in self._hash_keys:
                 # if existing in cache, get the index
-                index_ = key  # if using hash as cache keys, set the key
+                index = self._hash_keys.index(key)
 
-        if isinstance(index_, int) and index_ % len(self) >= self.cache_num:  # support negative index
+        if index % len(self) >= self.cache_num:  # support negative index
             # no cache for this index, execute all the transforms directly
-            return super()._transform(index_)
+            return super()._transform(index)
```

Comment:
it seems, at this point "index" could be in the hash space (if hashing by key), so this _transform(index) will run on the wrong element. see comment above for using index_hash

Reply:
Hi @myron ,

Maybe the logic here is a little bit confusing, actually, this line only executes when **the target item is not cached**, we don't change any cache content during training. So the `index` is certainly the original index instead of the hash list, here we call `super()` to treat it just as a regular `Dataset` item to run transforms.

Thanks.

Reply:
yes, you're correct that line happened to only execute when target item is not cached.  But this is very confusing,  and if someone changes anything e.g. self.cache_num calculation or it can lead to bugs that will be very hard to detect (it will return wrong image, without showing any errors)

Reply:
Sure, I updated all this part logic in the latest commit to make it more straightforward.

Thanks for your feedback.